packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+      this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.OPEN);
+      if (this.isQuickOpen_) {
+        this.adapter_.notifyClose();
+      } else {
+        this.closeAnimationEndTimerId_ = setTimeout(() => {
+          this.closeAnimationEndTimerId_ = 0;
+          this.adapter_.removeClass(MDCMenuSurfaceFoundation.cssClasses.ANIMATING_CLOSED);
+          this.adapter_.notifyClose();
+        }, numbers.TRANSITION_CLOSE_DURATION);
+      }
+    });
+
+    this.isOpen_ = false;
+    this.maybeRestoreFocus_();
+  }
+
+  /** Handle clicks and close if not within menu-surface element. */
+  handleBodyClick(evt: MouseEvent) {
+    const el = evt.target as Element;
+    if (this.adapter_.isElementInContainer(el)) {
+      return;
+    }
+    this.close();
+  }
+
+  /** Handle keys that close the surface. */
+  handleKeydown(evt: KeyboardEvent) {
+    const {keyCode, key, shiftKey} = evt;
+
+    const isEscape = key === 'Escape' || keyCode === 27;
+    const isTab = key === 'Tab' || keyCode === 9;
+
+    if (isEscape) {
+      this.close();
+    } else if (isTab) {
+      if (this.adapter_.isLastElementFocused() && !shiftKey) {
+        this.adapter_.focusFirstElement();
+        evt.preventDefault();
+      } else if (this.adapter_.isFirstElementFocused() && shiftKey) {
+        this.adapter_.focusLastElement();
+        evt.preventDefault();
+      }
+    }
+  }
+
+  private autoPosition_() {
+    // Compute measurements for autoposition methods reuse.
+    this.measurements_ = this.getAutoLayoutMeasurements_();
+
+    const corner = this.getOriginCorner_();
+    const maxMenuSurfaceHeight = this.getMenuSurfaceMaxHeight_(corner);
+    const verticalAlignment = (corner & CornerBit.BOTTOM) ? 'bottom' : 'top';
+    let horizontalAlignment = (corner & CornerBit.RIGHT) ? 'right' : 'left';
+    const horizontalOffset = this.getHorizontalOriginOffset_(corner);
+    const verticalOffset = this.getVerticalOriginOffset_(corner);
+    const {anchorSize, surfaceSize} = this.measurements_;
+
+    const position: Partial<MenuDistance> = {
+      [horizontalAlignment]: horizontalOffset ? horizontalOffset : 0,
+      [verticalAlignment]: verticalOffset ? verticalOffset : 0,
+    };
+
+    // Center align when anchor width is comparable or greater than menu surface, otherwise keep corner.
+    if (anchorSize.width / surfaceSize.width > numbers.ANCHOR_TO_MENU_SURFACE_WIDTH_RATIO) {
+      horizontalAlignment = 'center';
+    }
+
+    // If the menu-surface has been hoisted to the body, it's no longer relative to the anchor element
+    if (this.isHoistedElement_ || this.isFixedPosition_) {
+      this.adjustPositionForHoistedElement_(position);
+    }
+
+    this.adapter_.setTransformOrigin(`${horizontalAlignment} ${verticalAlignment}`);
+    this.adapter_.setPosition(position);
+    this.adapter_.setMaxHeight(maxMenuSurfaceHeight ? maxMenuSurfaceHeight + 'px' : '');
+  }
+
+  /**
+   * @return Measurements used to position menu surface popup.
```

Comment:
remove closure comments

Reply:
This is actually [TypeDoc](https://typedoc.org/guides/doccomments/) syntax 😀